pilot/pkg/networking/core/v1alpha3/httproute.go

```diff
@@ -463,7 +463,13 @@ func getVirtualHostsForSniffedServicePort(vhosts []*route.VirtualHost, routeName
 	}
 
 	if len(virtualHosts) == 0 {
-		virtualHosts = vhosts
+		return virtualHosts
+	} else if len(virtualHosts) == 1 {
+		virtualHosts[0].Domains = []string{"*"}
```

Comment:
I understand it happens when routename is in vhost.domains. And this could happen on the condition

```
if listenerProtocol == istionetworking.ListenerProtocolAuto &&
			sniffingEnabled && listenerOpts.bind != actualWildcard && listenerOpts.service != nil {
			rdsName = string(listenerOpts.service.ClusterLocal.Hostname) + ":" + strconv.Itoa(listenerOpts.port.Port)
		}
```

But i could not understand how this could fix the original issue

Reply:
Basically the original issue the route is bound to a listener on <vip>. Their request matches that vip, but the host header doesn't match any domains.
This makes the route match ALL host headers so it doesn't matter that they didn't have  a match

Reply:
You mean the original issue is the svc port is `auto-xxx` and for headless with it , we build a listener for each instance with name `podip:port`.